src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs

```diff
@@ -24,40 +24,27 @@ public override BoundNode VisitDeconstructionAssignmentOperator(BoundDeconstruct
             }
 
             BoundExpression loweredRight = VisitExpression(node.Right);
+            ImmutableArray<BoundExpression> rhsValues;
 
-            // prepare out parameters for Deconstruct
-            var deconstructParameters = node.DeconstructMember.Parameters;
-            var outParametersBuilder = ArrayBuilder<BoundExpression>.GetInstance(deconstructParameters.Length);
-            Debug.Assert(deconstructParameters.Length == node.LeftVariables.Length);
-
-            for (int i = 0; i < numVariables; i++)
+            // get or make right-hand-side values
+            if (node.Right.Type.IsTupleType)
             {
-                var localSymbol = new SynthesizedLocal(_factory.CurrentMethod, deconstructParameters[i].Type, SynthesizedLocalKind.LoweringTemp);
-
-                var localBound = new BoundLocal(syntax,
-                                                localSymbol,
-                                                null,
-                                                deconstructParameters[i].Type
-                                                ) { WasCompilerGenerated = true };
-
-                temps.Add(localSymbol);
-                outParametersBuilder.Add(localBound);
+                rhsValues = AccessTupleFields(node, loweredRight, temps, stores);
+            }
+            else
+            {
+                rhsValues = CallDeconstruct(node, loweredRight, temps, stores);
             }
 
-            var outParameters = outParametersBuilder.ToImmutableAndFree();
-
-            // invoke Deconstruct
-            var invokeDeconstruct = MakeCall(syntax, loweredRight, node.DeconstructMember, outParameters, node.DeconstructMember.ReturnType);
-            stores.Add(invokeDeconstruct);
-
-            // assign from out temps to lhs receivers
-            for (int i = 0; i < numVariables; i++)
+            // assign from rhs values to lhs receivers
+            int numAssignments = node.Assignments.Length;
+            for (int i = 0; i < numAssignments; i++)
             {
                 // lower the assignment and replace the placeholders for source and target in the process
                 var assignmentInfo = node.Assignments[i];
 
                 AddPlaceholderReplacement(assignmentInfo.LValuePlaceholder, lhsReceivers[i]);
-                AddPlaceholderReplacement(assignmentInfo.RValuePlaceholder, outParameters[i]);
+                AddPlaceholderReplacement(assignmentInfo.RValuePlaceholder, rhsValues[i]);
```

Comment:
Although the code above needs to be updated after LDM decision to not bundle conversion with assignment, I think it is otherwise correct and matches your description.

All the LHS side-effects have already been stacked on line 23: `lhsReceivers.Add(TransformCompoundAssignmentLHS(variable, stores, temps, isDynamicAssignment: false));`. That is step (1) in your description.

The loop that you commented on only takes care of conversion+assigment (step 2 from your description).